index.bs

```diff
@@ -124,22 +131,78 @@ A conversion report is a [=struct=] with the following items:
+<h3 algorithm id="parsing-conversion-destination">Parsing a conversion destination</h3>
 
-<h3 algorithm id="creating-impression">Creating an impression</h3>
-TODO
+To <dfn>parse a conversion destination</dfn> from an <{a}> tag |anchor|,
+1. Let |url| be the result of running the [=URL parser=] on the value of
+    the |anchor|'s <{a/conversiondestination}>.
+1. Return the result of [=obtain a site|obtaining a site=] from |url|'s
+    [=url/origin=].
+
+<h3 algorithm id="creating-impression">Activating an impression</h3>
+
+To <dfn>activate an impression</dfn> from an <{a}> tag |anchor|,
+1. Let |currentTime| be the current time.
+1. Let |impression| be a new [=impression=] struct whose items are:
+
+: [=impression/impression source=]
+:: |anchor|'s [=relevant settings object=]'s [=environment/top-level origin=].
+: [=impression/impression data=]
+:: The result of applying [=parsing conversion data=] to |anchor|'s
+    <{a/impressiondata}> attribute.
+: [=impression/conversion destination=]
+:: The result of running [=parse a conversion destination=] on |anchor|.
+: [=impression/reporting endpoint=]
+:: The [=url/origin=] of the result of running the [=URL parser=] on the value
+    of |anchor|'s <{a/reportingorigin}> attribute.
+: [=impression/expiry=]
+:: |currentTime| +  <{a/impressionexpiry}> milliseconds.
+: [=impression/impression time=]
+:: |currentTime|.
+
+Issue: Need to spec how to store the impression.
+
+<h3 algorithm id="creating-a-conversion">Creating a conversion</h3>
+
+To <dfn>create a conversion</dfn> from a [=url=] |url| and a
+    [=environment settings object=] |environment|.
+
+: [=conversion/conversion source=]
+:: |environment|'s [=environment/top-level origin=].
+: [=conversion/conversion data=]
+:: The result of applying [=parsing conversion data=] to the value associated with the
+    |conversion-data| field of |url|'s query part.
+: [=conversion/conversion time=]
+:: Be the current time.
+
+Issue: Formalize how to parse the query similar to URLSearchParams.
 
 <h3 algorithm id="register-conversion">Register a conversion</h3>
-TODO
 
-<h3 algorithm id="parsing-the-metadata">Parsing the metadata</h3>
+To <dfn>register a conversion</dfn> from a [=request=] |request|, run the following steps:
+
+1. If |request|'s [=request/current url's=] [=url/path=] is not `.well-known/register-conversion`,
+    return.
+1. If |request|'s [=request/redirect count=] is less than 1, return.
+1. Let |previousUrl| be the second to last [=URL=] in |request|'s
+    [=request/URL list=].
+1. If |request|'s [=request/current url's=] [=url/origin=] is not [=same origin=] with
+    |previousUrl|'s [=url/origin=], return.
+1. Let |conversionToRegister| be the result of applying [=create a conversion=] from the
+    request's [=request/current url=].
+
+Note: the restriction to require a redirect is needed to ensure that the
+request's origin is aware and in control of the conversion registration.
+
+Issue: Need to spec how to store the conversion.
```

Comment:
I'd probably put this into a list item as well, again to show readers that the algorithm isn't complete.

Reply:
Done